src/sagemaker/estimator.py

```diff
@@ -2644,9 +2646,15 @@ def set_hyperparameters(self, **kwargs):
         training code on SageMaker. For convenience, this accepts other types
         for keys and values, but ``str()`` will be called to convert them before
         training.
+
+        If a source directory is specified, this method escapes the dict argument as JSON,
+        and updates the private hyperparameter attribute.
         """
-        for k, v in kwargs.items():
-            self._hyperparameters[k] = v
+        if self.source_dir:
```

Comment:
is this a strict contract between `source_dir` and use of `_json_encode_hyperparameters`?

Reply:
yes. If we always encode the hyper parameters, some unit tests will fail.